csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -0,0 +1,154 @@
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Apache.Arrow.Ipc
+{
+    internal class ArrowStreamReaderImplementation : ArrowReaderImplementation
+    {
+        public Stream BaseStream { get; }
+        protected ArrayPool<byte> Buffers { get; }
+
+        public ArrowStreamReaderImplementation(Stream stream)
+        {
+            BaseStream = stream;
+            Buffers = ArrayPool<byte>.Create();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                BaseStream.Dispose();
+            }
+        }
+
+        public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationToken cancellationToken)
+        {
+            // TODO: Loop until a record batch is read.
+            cancellationToken.ThrowIfCancellationRequested();
+            return await ReadRecordBatchAsync(cancellationToken);
+        }
+
+        public override RecordBatch ReadNextRecordBatch()
+        {
+            throw new NotImplementedException();
+        }
+
+        protected async Task<RecordBatch> ReadRecordBatchAsync(CancellationToken cancellationToken = default)
+        {
+            await ReadSchemaAsync();
+
+            var bytesRead = 0;
+
+            byte[] lengthBuffer = null;
+            byte[] messageBuff = null;
+            byte[] bodyBuff = null;
+
+            try
+            {
+                // Get Length of record batch for message header.
+
+                lengthBuffer = Buffers.Rent(4);
+                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken);
+                var messageLength = BitConverter.ToInt32(lengthBuffer, 0);
+
+                if (messageLength == 0)
+                {
+                    //reached the end
+                    return null;
+                }
+
+                messageBuff = Buffers.Rent(messageLength);
+                bytesRead += await BaseStream.ReadAsync(messageBuff, 0, messageLength, cancellationToken);
+                var message = Flatbuf.Message.GetRootAsMessage(new FlatBuffers.ByteBuffer(messageBuff));
+
+                bodyBuff = Buffers.Rent((int)message.BodyLength);
+                var bodybb = new FlatBuffers.ByteBuffer(bodyBuff);
+                bytesRead += await BaseStream.ReadAsync(bodyBuff, 0, (int)message.BodyLength, cancellationToken);
+
+                return CreateArrowObjectFromMessage(message, bodybb);
```

Comment:
`message` here is created around a `ByteBuffer` that wraps `messageBuff`, and then presumably this "ArrowObject" will wrap that.  Does that end up making a copy of the original data, or does it store a reference to the original byte[]?  If the former, fine.  If the latter, this is dangerous, as we then proceed to return that buffer back to the pool.

Reply:
It makes a copy. See https://github.com/apache/arrow/pull/3736/files#diff-59f21081ce1a3587c235beb879c0150fR147 below in this same class:


```c#
        protected override ArrowBuffer CreateArrowBuffer(ReadOnlyMemory<byte> data)
        {
            // need to use the Buffer.Builder because we are currently renting the memory to
            // read messages
            return new ArrowBuffer.Builder<byte>(data.Length)
                .Append(data.Span)
                .Build();
        }
```

The extra copy will be addressed with [ARROW-4503: ArrowStreamReader allocates and copies data excessively](https://issues.apache.org/jira/browse/ARROW-4503)